### PR TITLE
Fix HPA conversion handling of annotations

### LINKED ISSUE #88738: Missing validation for HPA annotation
Recently, I tried to create an HPA object with an empty value for the `autoscaling.alpha.kubernetes.io/conditions` annotation and even though the response from apiserver is a 500, the object itself still got created (I verified this by querying for the key from the etcd). And since then `kubectl get/list hpa` calls are failing because apiserver seems to be unable to parse that particular object. Even worse, I wasn't even able to delete that object through the k8s api (had to delete the key directly from etcd).

This was on a v1.14.9 cluster.

Here's a minimal hpa object to reproduce the issue (both `v2beta1` and `v2beta2` seem to have it):
```
{
    "kind": "HorizontalPodAutoscaler",
    "apiVersion": "autoscaling/v2beta1",
    "metadata": {
        "name": "bad-hpa-object",
        "annotations": {
            "autoscaling.alpha.kubernetes.io/conditions": ""
        }
    },
    "spec": {
        "scaleTargetRef": {
            "kind": "Deployment",
            "name": "test",
            "apiVersion": "apps/v1"
        },
        "maxReplicas": 1
    }
}
```

Failing "create hpa" request:
```
$ kubectl create -f bad-hpa-object.json --v=8
...
I0302 18:45:37.912801   25603 request.go:942] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"unexpected end of JSON input","code":500}
I0302 18:45:37.913015   25603 helpers.go:196] server response object: [{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "error when creating \"hpa.json\": unexpected end of JSON input",
  "code": 500
}]
```

Kubectl call failing after above request:
```
$ kubectl get hpa
Error from server: unexpected end of JSON input
```

But object itself is present in etcd and I manually deleted it:
```
$ ETCDCTL_API=3 etcdctl get --keys-only --prefix "/registry/horizontalpodautoscalers/"
/registry/horizontalpodautoscalers/default/bad-hpa-object
$
$ etcdctl del /registry/horizontalpodautoscalers/default/bad-hpa-object
1
```

Which fixed the issue:
```
$ kubectl get hpa
No resources found.
```

cc @kubernetes/sig-api-machinery-bugs (I'm tagging apimachinery here because even though this could be a missing validation for a particular API owned by @kubernetes/sig-autoscaling-bugs , I'm wondering why apiserver created the object even though it returned a 500 to the client)

### PR DESCRIPTION
**Pull request by liggitt**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Fixes HPA handling of annotations in conversions:
* Ignores previously persisted malformed annotations
* Clears annotations used for round-tripping fields from other API versions after conversion

**Which issue(s) this PR fixes**:
Fixes #88738

**Does this PR introduce a user-facing change?**:
```release-note
Fixes conversion error for HorizontalPodAutoscaler objects with invalid annotations
```

/sig autoscaling
/cc @smarterclayton

### CODE REVIEW

File: pkg/apis/autoscaling/v1/conversion.go [hunk 1]
```diff
@@ -339,47 +356,40 @@ func Convert_v1_HorizontalPodAutoscaler_To_autoscaling_HorizontalPodAutoscaler(i
...
+		if err := json.Unmarshal([]byte(otherMetricsEnc), &otherMetrics); err == nil {
+			// the normal Spec conversion could have populated out.Spec.Metrics with a single element, so deal with that
+			outMetrics := make([]autoscaling.MetricSpec, len(otherMetrics)+len(out.Spec.Metrics))
+			for i, metric := range otherMetrics {
+				if err := Convert_v1_MetricSpec_To_autoscaling_MetricSpec(&metric, &outMetrics[i], s); err != nil {
+					return err
+				}
 			}
+			if out.Spec.Metrics != nil {
+				outMetrics[len(otherMetrics)] = out.Spec.Metrics[0]
```
Comment (smarterclayton) [inline pkg/apis/autoscaling/v1/conversion.go:368]: we are guaranteed that out.Spec.Metrics is non empty?  I guess your test covers that?  It's weird to assume out.Spec.Metrics has length 1 without a special guard to me.  Non obvious in code.

Comment (liggitt) [reply pkg/apis/autoscaling/v1/conversion.go:368]: this is unchanged, see https://github.com/kubernetes/kubernetes/pull/89963/files?w=1 for the changes without whitespace

Comment (liggitt) [reply pkg/apis/autoscaling/v1/conversion.go:368]: Convert_v1_HorizontalPodAutoscalerSpec_To_autoscaling_HorizontalPodAutoscalerSpec appears to populate out.Metrics with a single item list when TargetCPUUtilizationPercentage is non-nil. Will check what happens when it is nil.

Comment (liggitt) [reply pkg/apis/autoscaling/v1/conversion.go:368]: yeah, `out.Spec.Metrics` is nil unless populated by `Convert_v1_HorizontalPodAutoscalerSpec_To_autoscaling_HorizontalPodAutoscalerSpec`

Comment (smarterclayton) [reply pkg/apis/autoscaling/v1/conversion.go:368]: ok.

**Comment by smarterclayton** [comment]
/lgtm
/approve

God help us

**Pull request merged**